src/main/java/com/linkedin/kmf/apps/MirrorPipelineMonitor.java

```diff
@@ -0,0 +1,448 @@
+import com.linkedin.kmf.services.DefaultMetricsReporterService;
+import com.linkedin.kmf.services.JettyService;
+import com.linkedin.kmf.services.JolokiaService;
+import com.linkedin.kmf.services.ProduceService;
+import com.linkedin.kmf.services.TopicManagementService;
+import com.linkedin.kmf.services.configs.CommonServiceConfig;
+import com.linkedin.kmf.services.configs.ConsumeServiceConfig;
+import com.linkedin.kmf.services.configs.DefaultMetricsReporterServiceConfig;
+import com.linkedin.kmf.services.configs.ProduceServiceConfig;
+import com.linkedin.kmf.services.configs.TopicManagementServiceConfig;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.sourceforge.argparse4j.ArgumentParsers;
+import net.sourceforge.argparse4j.inf.ArgumentParser;
+import net.sourceforge.argparse4j.inf.Namespace;
+import org.apache.kafka.common.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static net.sourceforge.argparse4j.impl.Arguments.store;
+
+/*
+ * The MirrorPipelineMonitor app is intended to monitor the health and performance of kafka brokers across a mirror maker
+ * pipeline. It creates one producer and one consumer with the given configuration, produces messages with increasing
+ * integer in the message String, consumes messages, and keeps track of number of lost messages, duplicate messages, e2e
+ * latency, throughput, etc. As well, MirrorPipelineMonitor creates a topic management service for every kafka and zookeeper
+ * broker in the pipeline.
+ *
+ * MirrorPipelineMonitor app exports these metrics via JMX. It also periodically report metrics if DEBUG level logging
+ * is enabled. This information can be used by other application to trigger alert when kafka brokers fail. It also
+ * allows users to track pipeline performance, e.g. latency and throughput.
+ */
+
+public class MirrorPipelineMonitor implements App {
+  private static final Logger LOG = LoggerFactory.getLogger(MirrorPipelineMonitor.class);
+
+  private final ProduceService _produceService;
+  private final ConsumeService _consumeService;
+  private final List<TopicManagementService> _topicManagementServices;
+  private final String _name;
+
+  public MirrorPipelineMonitor(Map<String, Object> props, String name) throws Exception {
+    _name = name;
+    _produceService = new ProduceService(createProduceServiceConfig(props), name);
+    _consumeService = new ConsumeService(createConsumeServiceConfig(props), name);
+    _topicManagementServices = new ArrayList<>();
+    createTopicManagementServices(props, name);
+  }
+
+  private void createTopicManagementServices(Map<String, Object> props, String name) {
+    Map<String, Object> topicManagementConfig = createTopicManagementServiceConfig(props);
+    Double partitionsToBrokerRatio;
+    if (topicManagementConfig.containsKey(TopicManagementServiceConfig.PARTITIONS_TO_BROKER_RATIO_CONFIG)) {
```

Comment:
You don't need to read `TopicManagementServiceConfig.PARTITIONS_TO_BROKER_RATIO_DEFAULT` directly. `TopicManagementServiceConfig` handled these for you. Actually, that is the purpose of having `TopicManagementServiceConfig`.

You can probably learn how to use `TopicManagementServiceConfig` by looking at `TopicManagementService`. I think the code below may be simpler:

```
Map<String, Object> topicMangementProps = (Map<String, Object>) props.getOrDefault(MirrorPipelineMonitorConfig.TOPIC_MANAGEMENT_SERVICE_CONFIG, new HashMap<String, Object>());
topicMangementProps.put(CommonServiceConfig.TOPIC_CONFIG, props.get(CommonServiceConfig.TOPIC_CONFIG));
TopicManagementServiceConfig config = new TopicManagementServiceConfig(topicMangementProps);
double partitionsToBrokerRatio = config.getDouble(TopicManagementServiceConfig.PARTITIONS_TO_BROKER_RATIO_THRESHOLD);
``